src/qibolab/instruments/rfsoc.py

```diff
@@ -1,54 +1,164 @@
 """ RFSoC FPGA driver.
 
-This driver needs the library Qick installed
-Supports the following FPGA:
+This driver needs a Qibosoq server installed and running
+Tested on the following FPGA:
  *   RFSoC 4x2
+ *   ZCU111
 """
 
-import pickle
+import json
 import socket
-from dataclasses import dataclass
-from typing import Dict, List, Optional, Tuple, Union
+from dataclasses import asdict
+from typing import Dict, List, Tuple, Union
 
 import numpy as np
+import qibosoq.components as rfsoc
 
 from qibolab import AcquisitionType, AveragingMode, ExecutionParameters
 from qibolab.instruments.abstract import AbstractInstrument
 from qibolab.platform import Qubit
-from qibolab.pulses import PulseSequence, PulseType
+from qibolab.pulses import Drag, Gaussian, Pulse, PulseSequence, PulseType, Rectangular
 from qibolab.result import IntegratedResults, SampleResults
 from qibolab.sweeper import Parameter, Sweeper
 
+HZ_TO_MHZ = 1e-6
+NS_TO_US = 1e-3
+
+
+def convert_qubit(qubit: Qubit) -> rfsoc.Qubit:
+    """Convert `qibolab.platforms.abstract.Qubit` to `qibosoq.abstract.Qubit`"""
+    if qubit.flux:
+        return rfsoc.Qubit(qubit.flux.bias, qubit.flux.ports[0][1])
+    return rfsoc.Qubit(0.0, None)
+
+
+def set_rfsoc_pulse_shape(pulse: Pulse, rfsoc_pulse: rfsoc.Pulse):
+    """Set pulse shape parameters in rfsoc pulse object"""
+
+    rfsoc_pulse.shape = pulse.shape.name.lower()
+    if rfsoc_pulse.shape in {"gaussian", "drag"}:
+        rfsoc_pulse.rel_sigma = pulse.shape.rel_sigma
+        if rfsoc_pulse.shape == "drag":
+            rfsoc_pulse.beta = pulse.shape.beta
```

Comment:
Unless you have performance limitations (and this does not seem to be the case) better to keep the input object untouched, and return a new one (and leave memory management up to the garbage collector).
```suggestion
def rfsoc_pulse_shape(pulse: Pulse, rfsoc_pulse: rfsoc.Pulse):
    """Set pulse shape parameters in rfsoc pulse object"""
    new = copy.copy(rfsoc_pulse)

    shape = new.shape = pulse.shape.name.lower()
    if shape in {"gaussian", "drag"}:
        new.rel_sigma = pulse.shape.rel_sigma
        if shape == "drag":
            new.beta = pulse.shape.beta

    return new
```